controllers/hotrod_upgrade_controller.go

```diff
@@ -0,0 +1,702 @@
+	kube "github.com/infinispan/infinispan-operator/pkg/kubernetes"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"regexp"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"strconv"
+	"strings"
+)
+
+type HotRodRollingUpgradeReconciler struct {
+	client.Client
+	scheme     *runtime.Scheme
+	log        logr.Logger
+	kubernetes *kube.Kubernetes
+	eventRec   record.EventRecorder
+}
+
+type HotRodRollingUpgradeRequest struct {
+	*HotRodRollingUpgradeReconciler
+	ctx                context.Context
+	infinispan         *ispnv1.Infinispan
+	reqLogger          logr.Logger
+	cluster            infinispan.ClusterInterface
+	currentStatefulSet string
+}
+
+var nameRegexp = regexp.MustCompile(`(.*)-([\d]+$)`)
+
+func (r *HotRodRollingUpgradeReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	name := "hotrod-rolling-upgrade"
+
+	r.Client = mgr.GetClient()
+	r.log = ctrl.Log.WithName("controllers").WithName(strings.Title(name))
+	r.scheme = mgr.GetScheme()
+	r.kubernetes = kube.NewKubernetesFromController(mgr)
+	r.eventRec = mgr.GetEventRecorderFor(name + "-controller")
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&ispnv1.Infinispan{}).
+		WithEventFilter(predicate.Funcs{
+			UpdateFunc: func(updateEvent event.UpdateEvent) bool {
+				return true
+			},
+		}).
+		Complete(r)
+}
+
+func (r *HotRodRollingUpgradeReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := r.log.WithValues("Reconciling", "HotRodRollingUpgrade", "Request.Namespace", request.Namespace, "Request.Name", request.Name)
+
+	// Get associated infinispan resource
+	ispn := &ispnv1.Infinispan{}
+	if err := r.Get(ctx, request.NamespacedName, ispn); err != nil {
+		if kerrors.IsNotFound(err) {
+			reqLogger.Info("Infinispan CR not found")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("unable to fetch Infinispan CR %w", err)
+	}
+
+	// Check if the CR is configured to do Hot Rod rolling upgrades
+	if ispn.Spec.Upgrades != nil && ispn.Spec.Upgrades.Type != ispnv1.UpgradeTypeHotRodRolling {
+		return reconcile.Result{}, nil
+	}
+
+	// Skip for x-site
+	if ispn.HasSites() {
+		return reconcile.Result{}, errors.New("hot rod rolling upgrades not supported for clusters with X-Site")
+	}
+
+	// Skip for non-datagrid
+	if ispn.Spec.Service.Type != ispnv1.ServiceTypeDataGrid {
+		return reconcile.Result{}, errors.New("hot rod rolling upgrades only supported for ServiceTypeDataGrid")
```

Comment:
I see in the CI logs this is being output several times in tests that don't use `HotRodRolling`: https://ci.infinispan.org/job/Infinispan%20Operator/job/PR-1260/103/consoleFull

Reply:
that's weird, looking...

Reply:
It is a race condition where the upgrade field is still ```nil``` and ```ApplyDefaults()``` hasn't been called yet. Will fix it